### ISSUE TITLE
关于复现 DR-BERT 结果没论文里好的问题

### ISSUE THREAD

**Issue by wulaoshi** [gmftbyGMFTBY/SimpleReDial-v1#2]
作者你好，非常感谢你的工作。我最近在复现 dense retrieval 发现结果不如论文里的号。
具体配置：加载哈工大的 roberta，batch_size 64，epochs_num 5， grad_clip 5.0，learning_rate 5e-5，max_len 分别为 256, 64；使用 ECD 数据，测试集上所有数值均比论文里低一些，请问有什么纰漏吗？

**Comment by gmftbyGMFTBY** [comment]
直接使用roberta或者bert效果是不行的（为了和之前的pretrained方法进行公平比较，我不用roberta，用的是标准的bert-base-uncased和bert-base-chinese），你尝试了domain adaptive pretraining了么，我再released的checkpoint里面放出来了我在这几个数据集上做的domain adaptive pretraining的模型参数，直接加载那些参数再做训练会好很多。


| |
GMFTBY
|
|
***@***.***
|
签名由网易邮箱大师定制
在2021年12月1日 ***@***.***> 写道：

作者你好，非常感谢你的工作。我最近在复现 dense retrieval 发现结果不如论文里的号。
具体配置：加载哈工大的 roberta，batch_size 64，epochs_num 5， grad_clip 5.0，learning_rate 5e-5，max_len 分别为 256, 64；使用 ECD 数据，测试集上所有数值均比论文里低一些，请问有什么纰漏吗？

—
You are receiving this because you are subscribed to this thread.
Reply to this email directly, view it on GitHub, or unsubscribe.
Triage notifications on the go with GitHub Mobile for iOS or Android.

**Comment by gmftbyGMFTBY** [comment]
如果你要再别的数据集上训练的话，建议先运行一下下面这个命令进行domain adaptive pretraining（本质上就是单句的roberta-style pretraining）:


./scripts/train.sh <dataset_name> bert-fp-mono <cuda_ids>


另外你需要检查一下我代码里面的dataloader类里面的代码，因为我没有在ECD数据集上跑过实验，你需要先检查一下是否会出现什么问题。


| |
GMFTBY
|
|
***@***.***
|
签名由网易邮箱大师定制
在2021年12月1日 ***@***.***> 写道：

作者你好，非常感谢你的工作。我最近在复现 dense retrieval 发现结果不如论文里的号。
具体配置：加载哈工大的 roberta，batch_size 64，epochs_num 5， grad_clip 5.0，learning_rate 5e-5，max_len 分别为 256, 64；使用 ECD 数据，测试集上所有数值均比论文里低一些，请问有什么纰漏吗？

—
You are receiving this because you are subscribed to this thread.
Reply to this email directly, view it on GitHub, or unsubscribe.
Triage notifications on the go with GitHub Mobile for iOS or Android.

**Comment by gmftbyGMFTBY** [comment]
@wulaoshi 你有ECD数据集的链接之类的信息么，我可以把这个数据集的实验补充一下（还是说ECD是E-commerce dataset?）如果ECD数据集是E-commerce dataset的话，你试一下我在released checkpoint链接里面的E-commerce模型的参数：
1. epoch: 10
2. max_length: 256/64
3. batch_size: 64
4. full_turn_length: 9
5. learning_rate: 5e-5
6. grad_clip: 5.0
7. warmup_step: 0.0
8. bert-fp-mono checkpoint

**Comment by wulaoshi** [comment]
> @wulaoshi 你有ECD数据集的链接之类的信息么，我可以把这个数据集的实验补充一下（还是说ECD是E-commerce dataset?）如果ECD数据集是E-commerce dataset的话，你试一下我在released checkpoint链接里面的E-commerce模型的参数：
> 
> 1. epoch: 10
> 2. max_length: 256/64
> 3. batch_size: 64
> 4. full_turn_length: 9
> 5. learning_rate: 5e-5
> 6. grad_clip: 5.0
> 7. warmup_step: 0.0
> 8. bert-fp-mono checkpoint

谢谢你的回复。目前我还没直接使用你的代码，使用自己实现的模型，现阶段是想与你的指标对齐，以便检查否有其它问题。ECD数据是直接使用的官方开源的，进行了简单清洗替换与添加[USER]，[SYS] 等特殊token，模型代码大概这样：
![image](https://user-images.githubusercontent.com/27938964/144162359-7f079eef-1161-498a-a7c8-3452b16ef7d2.png)
测试时batch_size也是64，与你公布的 (一条一条测) 不一致，不知是否是这个原因，我先多找下异同。再次感谢你的回复。

**Comment by wulaoshi** [comment]
另外，不知道请问你在训练的时，5个epoch后，loss 一般降为多少？

**Comment by wulaoshi** [comment]
另外，请问你在训练时，(DR-BERT w/o. FG)  5个 epoch 后 loss 一般降为多少？谢谢。

**Comment by gmftbyGMFTBY** [comment]
我看了一下大概有这么几个问题：
1. 你用的是InfoNCE loss来做的，也就是说你算的是带temperature的cosine similarity，这个temperature你最好设定为0.07或者一个比较小的数值，否则效果很差（因为temperature太大不能有效在语义空间中推开样本之间距离），我发现实际运行的时候直接用dot product similarity效果就比较稳定，所以我没有用temperature-based cosine similarity
2. 如果是batch测和单条测试是没有什么差别的，问题不在这里
3. 直接用的BERT模型参数来做效果不太行，必须要做domain adaptive pretraining（在下游这些response selection数据集上继续用NSP和MLM训练来预热模型参数），之前的工作比如SA-BERT, BERT-FP都用到了domain adaptive pretraining,你也可以直接用他们released的domain adaptive pretraining模型参数（例如BERT-FP：https://github.com/hanjanghoon/BERT_FP/tree/master/； 或者我released的：https://drive.google.com/drive/folders/1y48ky8twFKbvcu9TCJ9DxBG9lMTYnRYI  )
4. E-commerce数据集要用10个epoch，5个还没有完全收敛，我用了8卡并行跑的，所以收敛快
5. 关于loss的问题，我训练了10个epoch的tensorboard curve如下所示，可以跑到0.1464：
![image](https://user-images.githubusercontent.com/27548710/144163729-84423a1c-2069-44b9-9ab3-56206d7524e4.png)

**Comment by wulaoshi** [comment]
谢谢，应该就是你上面提到的。我5个epoch 后 loss为 0.47，并未收敛。我再根据你的回复重新训练，看下结果。谢谢！~
